lib/common/xml.c

```diff
@@ -1489,15 +1487,16 @@ pcmk__xe_log(int log_level, const char *file, const char *function, int line,
     name = crm_element_name(data);
 
     if (pcmk_is_set(options, xml_log_option_open)) {
-        char *buffer = NULL;
+        GString *buffer = g_string_sized_new(128);
```

Comment:
This is interesting. Some lines (e.g. a typical lrm_rsc_op) will be far longer, 512 or even bigger would be a good start for them. On the other hand most lines are likely to be smaller than 128.

I'm thinking the caller should allocate the string and pass it in, so we're only doing a single memory allocation and reusing it for each element. Probably the best approach would be to keep pcmk__xe_log() as a wrapper that allocates a large string and calls the actual function, to avoid duplication at the callers. That might even boost performance noticeably.

Regarding the initial sizes in general throughout, note that buffer_print() used a chunk size of 1024. Something between 512 and 1024 is probably about right generically, though smaller could be appropriate in some cases.

Reply:
> Probably the best approach would be to keep pcmk__xe_log() as a wrapper that allocates a large string and calls the actual function, to avoid duplication at the callers.

That sounds like a good idea. Allocate one buffer in advance, and truncate where we're currently allocating. Truncation is lightweight for GStrings.

> Regarding the initial sizes in general throughout, note that buffer_print() used a chunk size of 1024. Something between 512 and 1024 is probably about right generically

True. The extra overhead for the initial allocation is probably minimal at that scale, and allocating a larger amount reduces the likelihood of needing to reallocate. I'll try using 4096 in `digest.c` (since it looks like we expect to handle the whole tree) and 1024 elsewhere.

Reply:
I thought I bumped the buffers in `pcmk__xml_log()` to 1024 too. I won't change it now since a later commit replaces both of these calls with a single one in a wrapper.